corgi/core/models.py

```diff
@@ -904,75 +910,84 @@ def latest_components(
         self,
         include: bool = True,
     ) -> "ComponentQuerySet":
-        """Return components from latest builds."""
-        names = self.values_list("name", flat=True).order_by().distinct().iterator()
+        """Return components from latest builds across all product streams."""
+        distinct_components = (
+            self.values_list("namespace", "name", "arch").order_by().distinct().iterator()
+        )
```

Comment:
There's no restriction on type, so if I include a name in the filter such as 'cloud.google.com/go/monitor', I get results such as 'pkg:golang/cloud.google.com/go/monitoring@v1.1.0' which uses semantic versioning, not RPM versioning, so the results could be incorrect.

```suggestion
        # This should probably go in constants.py
        RPM_SCHEMANTIC_CONDITION = Q(type="RPM") | Q(type="OCI", namespace="REDHAT")
        distinct_components = (
            self.filter(RPM_SCHEMANTIC_CONDITION).values_list("namespace", "name", "arch").order_by().distinct().iterator()
        )
```

Reply:
If we're comparing e.g. two Golang components, the epoch should always be "0" and the release should always be an empty "" string, right? In that case, we'd compare the two components using only their version strings.

From what I remember of the two standards, I think the version comparison logic for a semver is simpler than the version comparison logic for an RPM version. I'm happy to add a test for this case, but I'm 99% sure that the filter / the library we use will give the correct results for semantic versions.

The user request was to see the latest version of all components in a stream, not just the latest version of RPM / container components, so I don't want to limit the results we return unless necessary. I'd also rather keep the scope of this ticket focused on fixing this particular bug (where we don't split results by arch or namespace). I don't want to change the logic of what the filter returns until we're sure there's a bug.

Happy to discuss in Gchat or on a call if you'd like - please just let me know.

Reply:
Discussed on a call. We probably need more complex logic to support this usecase. We could have different version-comparison logic for different component types (which use different version types, e.g. semver).

For now we'll only use the latest filter on RPMs and REDHAT containers, which we know use RPM-style versions. Containers used by managed services (and stored in Quay) may be of type UPSTREAM if a service just uses an upstream container directly without repackaging / rebuilding it in Brew / the REDHAT namespace. So we also need to filter on ```Q(namespace=REDHAT, type=OCI)```.